### PR TITLE
[WIP]add smooth l1 forward kernel

### PR DESCRIPTION
**Pull request by yubofredwang**
## Summary
Add smooth l1 forward kernel. This loss is used during EAGLE training. See: https://github.com/SafeAILab/EAGLE/blob/a0e1e2c08604f0d215c531d9655b7ebacedde130/eagle/train/main.py#L231

We can achieve 5x speed up and 50% peak memory reduction.

Task left:
- [ ] Implement gradient calculations
- [ ] Add backward pass
- [ ] Add integration testing.

## Testing Done
Initial benchmark result is very promising:

```
Memory Usage Profiling
==================================================
PyTorch Memory: 0.00 MB
Triton Memory:  0.00 MB
Memory Ratio:   0.25x
jobuser [ /shared/user/Liger-Kernel/src/liger_kernel/ops ]$ python smooth_l1.py 
Running benchmarks on: cuda
================================================================================
Config          Reduction  PyTorch (ms)    Triton (ms)     Speedup   
------------------------------------------------------------------------------------------------------------------------
1024x131072     none       2.103           0.452           4.65      
2048x131072     none       4.123           0.823           5.01      
4096x131072     none       8.171           1.573           5.20      
8192x131072     none       16.300          3.073           5.31      
16384x131072    none       32.543          6.062           5.37      
========================================================================================================================
```

Benchmark completed!
Notes:
- Times are averaged over 50 iterations
- Max Diff: Maximum absolute difference between reference and triton
- Rel Error: Maximum relative error
- Speedup: PyTorch time / Triton time

Peak Memory Usage Profiling
==================================================
PyTorch Peak Memory: 168.01 MB
Triton Peak Memory:  72.00 MB
Memory Ratio:        2.33x

- Hardware Type: <BLANK>
- [ ] run `make test` to ensure correctness
- [ ] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by helloworld1** [comment]
Good job! Surprised this is faster than the C++ implementation in torch here https://github.com/pytorch/pytorch/blob/a4ec381302f8acd279033707b182bed30ffd2091/torch/csrc/api/include/torch/nn/functional/loss.h#L337

Wonder if we can compare a python implementation but with torch.compile.

**Comment by yubofredwang** [comment]
> Good job! Surprised this is faster than the C++ implementation in torch here https://github.com/pytorch/pytorch/blob/a4ec381302f8acd279033707b182bed30ffd2091/torch/csrc/api/include/torch/nn/functional/loss.h#L337
> 
> Wonder if we can compare a python implementation but with torch.compile.

Thanks for the feedback, here is a comparison with forward(excluding gradient calculations for backward):
```
Running benchmarks on: cuda
================================================================================
Config          Reduction  PyTorch (ms)    Triton (ms)     Speedup   
------------------------------------------------------------------------------------------------------------------------
1024x131072     none       0.563           0.445           1.27      
2048x131072     none       0.986           0.824           1.20      
4096x131072     none       1.896           1.571           1.21      
8192x131072     none       3.719           3.066           1.21      
16384x131072    none       9.565           6.058           1.58      
========================================================================================================================

Benchmark completed!
Notes:
- Times are averaged over 50 iterations
- Max Diff: Maximum absolute difference between reference and triton
- Rel Error: Maximum relative error
- Speedup: PyTorch time / Triton time

Peak Memory Usage Profiling
==================================================
PyTorch Peak Memory: 72.01 MB
Triton Peak Memory:  72.00 MB
Memory Ratio:        1.00x
```
The improvement was mainly because for LLM use cases. We need to do a batchmean reduction. In torch it would be more operations:

see how forward_full is implemented for reference.

I will add the backward today and make the PR official

**Comment by yubofredwang** [comment]
closing as more details benchmarking indicates the performance improvement is minimal
